### ISSUE TITLE
BUG?: Token output has 0 amount

### ISSUE THREAD

**Issue by ashisherc** [IntersectMBO/cardano-ledger#2444]
I have noticed some transactions on public testnet are having token transfers with 2 outputs. one output with token and amount and another one with same token but 0 amount.

Here's such tx on testnet,
https://testnet.cardanoscan.io/transaction/7666d0b34a040f757c085b39ffca90ee83a357b4d002746bec85280fbebb5c3d

And here's the CBOR of the block containing this tx,
```
820484828f1a002be4991a02271b3d58205e0f10a00658118ced084538c911994787ce326a5999f91dcf9e1689275b1d9b58205fe9977d3df5b45e0ebeca2a95e1afdffe168ea0e311ca196cd74554effe8dc85820d2caff8a74e8cd6151b95d08272c8f974ac069d9e07adca90909e78aa091ffc7825840ed585ee6a73c489a2728e2453f4cbb3fbe96c67f90608dd184fcebf6e00cb345b0ad85e342c374ef66fa6089f264b09c5f100b0bd5556c555d820a4e0394e1c858501d1863e1ae7f128b59b55a0bb507f46dee69902a9565d56268212165d602c9d9f7a670409f2d32bd1211dad7c17d91c978f61d3e77191a1ee5f9f76045e9f39090b5b65f1e58b866a3f1f61709d41f048258400013cd1f64bb858860a09f96feb6c6d8deb749404a118fbfc1841fd7992b69fc397e4366a1cb7bba10e8a20a1c6c63a924ceb0d425451e6ddfbe5d7aa871cb59585006aaa46f21ec61875ca7879cf16ea5032a38a23082ed65a7e7a26095df1d7c3ec8a77510deaa917daf231218e7c8d014b9612b9b7a2737452db159539a2c68aab0c28807bd2359aeea52e57eccf33b0a1902cb5820dcb2b6ea9dcb04e554e8ba4d888c12c9097a2b5e70bd01e557113b63db695c0858207fa60ca127662d8adf1b8e86f5ac433b324869088aff9babed63438b87bc47b1041901125840dcd6dd6b99d277934cebfed20418c9a52b10e0b1d6497c61c49b0e1a096559982b8b82846103ce36c12c8824307e5b3a011ca4adb6d1e29fa98cb3c790eb2b0705005901c08a376b4f14632a94a66fea59983170d98a6501285324fa8707e7fb24757eb6a44ebdedc376769ab24fd2d0119fc0a926ec5e949c6f8616116a5dd0935f69160c5632806e48fe9aeb49dd416e0eaad25041b02f4ab80dfb219d60248a74b2121c6abe5bf40f44444106bf768a433b17c779f803f127dd196c62f38c5010ff4a5d11d545ae1d78792672c1452aca9cbf25f0b29168bbb5bc5c1a3b9e594d9bd35d50c03cb4e4e501e7923061eaeb3f1a0f377c265aa9dce01b6c36a4db1c7ad59ee56c28cafb645ede88710e5a045f91151e40326bf3afbd81d2dc13ca9752d2c0d0575e2282b04c7134783ec103d2dbf9216cfbb17c7e8d40d690d2c42476229307a028c9e666b9704ee275e8f0ff26d816a1087ab24feb03fb455b472c7ded948088dcf57fd8a837abaf9522e2a430eee43ac16d3f9518130aa5042b7da30555e67eb92b1083f167d4bb98bead74db464088f3176d62605ffe1beda875c9ba055d1f118fb52953a9f87b276fdc830f553fb7efaf4713a252cf1319fb2e7c7d2c426bcca5726a471f65bbe3ac6b290c2f91e3c8cad1eaf8c2b09aa90e63b739ed29e6c1ce87911504e1122bc3810e83d34636671c13fc0cf739ae7c94115e857e81a500828258206e20ab451a9fdca29c52a53b105f3af0d15e857b1e5ecbe8b9659e8a5c8bd8de018258209bd3240ecc4baeabec7006e5eb7875d11368729ff8c60568a299df0ade9f9d18000182825839000205f8480d5ed699be1352116406c8da63d86c8903a3aeb69b7ffdbb4b78af166568d94edfcdb5a3fbb2bde76b72102c9579b6d49c9e8f76821a00169b08a1581c4247a1a0579972989f657c53bf6af7bfbbc459a170b45ecb61a719b7a14c6b6172616e206e61747572650182583900397dcee9b395f1964e4a28ecb2195b2a26ac958fe75bbdff9af16ec0e227f7a0725901adf4ee51e9250640b630c80cf5340780a070979258821a05bffdf1a1581c4247a1a0579972989f657c53bf6af7bfbbc459a170b45ecb61a719b7a14c6b6172616e206e617475726500021a0002d955031a022722fd0758204008a0aa62e06eb3d98483678da8a00029504525849b1d65953f8b3e7680e72481a10081825820add8b2e24b269281de65b9d6b79929413161bdb81a5d188888f840f5a4d0ba6c58408b2d5e036186c82bef81d8872ca498fedef5e8b60c17a36e63505e8554d57c62bc1526ab4a0d3d2711932f08220d0c5bcae670b0c6b1b1acf451f38dcd87a207a100a11902d1a178383432343761316130353739393732393839663635376335336266366166376266626263343539613137306234356563623631613731396237a16b6b6172616e6e6174757265a565696d6167657834697066733a2f516d61735762633855434b6f7234447344394375446872524d78386e4c507a375671706e55786875414a48397662637372637834697066733a2f516d64453548776d34786338674c6d67625858346e5862534e636e7571416d78624261686e4d5131674266537a55646e616d656c6b6172616e206e61747572656b6465736372697074696f6e716b6172616e206e6174757265206465736364747970656a696d6167652f6a706567
```

**Comment by JaredCorduan** [comment]
Thank you @ashisherc ! The spec does not forbid zero-valued tokens, but the implementation does enforce a canonical form (no zero-values in the map). It would therefore be a bug if this token showed up in the ledger state, but it does not:

```
~/cardano-node$ cardano-cli query utxo --testnet-magic 1097911063 --address addr_test1qquhmnhfkw2lr9jwfg5wevsetv4zdty43ln4h00lntckas8zylm6qujeqxklfmj3ayjsvs9kxryqeaf5q7q2quyhjfvqegfhlg
                           TxHash                                 TxIx        Amount
--------------------------------------------------------------------------------------
248b3c54c4198a6772c5e83b1b656f2341097c2955198a969c855e08588a9078     0        94796492 lovelace + TxOutDatumHashNone
248b3c54c4198a6772c5e83b1b656f2341097c2955198a969c855e08588a9078     1        1481480 lovelace + 1 0ccd37ee1dddf3ac6a807c0d2f9941099d05845d42aae07c560866b5.karan33333 + TxOutDatumHashNone
43bb9ab7827089171355b34792e79dfe595b4233bc0cef43cea7c6b8ab221abf     0        9000000 lovelace + TxOutDatumHashNone
65f9b52c0df00f9e118bf42679d5c9c31a0c3f4108a6c54e5f4b254d2ec379da     0        10715494 lovelace + 1 00a63f2cd08fb4cec077b47d845c28db9e0a0648d51898534e5a2cd4.karan nft + TxOutDatumHashNone
9c86c8170c4f8ba27542074d685acada266e1ae5fb8c203cd7e7e2b0aaa8f868     0        9000000 lovelace + TxOutDatumHashNone
ab2fdfa2b0110d693bf2a22055690a6a305fe4d64a091c079ca09b1a4c75e681     0        9000000 lovelace + TxOutDatumHashNone
ebe6dd110bd10c0d2dc831c0ab41f22a6410fa6ed8befb02a8a8004b13333fd8     1        8241054 lovelace + TxOutDatumHashNone

```

We do permit zero-valued token in the serialization, so all is well. But thanks for reporting!

**Comment by ashisherc** [comment]
Awesome! thank you!

**Comment by JaredCorduan** [comment]
Though not necessarily a bug, I think we might want to consider this a wart worth cleaning up in the future.

**Comment by ashisherc** [comment]
Thanks @JaredCorduan !

**Comment by JaredCorduan** [comment]
I just realized that #3240 is a duplicate of this issue. and is fixed by #3241.
